### PR TITLE
qdevices: Add daemon alive check in start_daemon()

### PR DESCRIPTION
**Pull request by zhenyzha**
Add daemon alive check in start_daemon()

ID:2052864
Signed-off-by: Zhenyu Zhang <zhenyzha@redhat.com>

### CODE REVIEW

**Comment by zhenyzha** [comment]
**normal test:**
`(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.arm64-pci: PASS (66.37 s)`

**’None -d’**
`(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.aarch64.io-github-autotest-qemu.tpm_verify_device.with_emulator.arm64-pci: PASS (44.53 s)`
**‘ -d’**
` (1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.aarch64.io-github-autotest-qemu.tpm_verify_device.with_emulator.arm64-pci: PASS (44.74 s)`

**negative test:**
`(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.arm64-pci: ERROR: VM 'avocado-vt-vm1' failed to start: Error occurred while executing make_create_command(). Check the log for traceback. (4.84 s)`
virttest.qemu_devices.utils.DeviceError: Failed to run virtiofs daemon: /bin/bash: line 1: /usr/libexec/virtiofsd: No such file or directory

`(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.aarch64.io-github-autotest-qemu.tpm_verify_device.with_emulator.arm64-pci: ERROR: VM 'avocado-vt-vm1' failed to start: Error occurred while executing make_create_command(). Check the log for traceback. (5.01 s)`
virttest.qemu_devices.utils.DeviceError: Failed to run swtpm daemon: /tmp/aexpect_j4yGnfDy/aexpect-q_5xtg6j.sh: line 1: /usr/bin/swtpm: No such file or directory

@YongxueHong @qcheng-redhat @xiagao 
Could you pls help review when you are free. thanks.

**Comment by qcheng-redhat** [comment]
Checked tpm test cases:

 (1/4) Host_RHEL.m8.u6.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: STARTED
\ (1/4) Host_RHEL.m8.u6.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: PASS (129.29 s)
 (2/4) Host_RHEL.m8.u6.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: STARTED
| (2/4) Host_RHEL.m8.u6.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: PASS (84.30 s)
 (3/4) Host_RHEL.m8.u6.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: STARTED
| (3/4) Host_RHEL.m8.u6.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: PASS (151.46 s)
 (4/4) Host_RHEL.m8.u6.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: STARTED
- (4/4) Host_RHEL.m8.u6.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: PASS (234.20 s)

For negative scenario:
(1/2) Host_RHEL.m8.u6.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: ERROR: VM 'avocado-vt-vm1' failed to start: Error occurred while executing make_create_command(). Check the log for traceback. 

Log:
01:32:57 ERROR|     raise DeviceError("Failed to run swtpm daemon: %s" % output)
01:32:57 ERROR| virttest.qemu_devices.utils.DeviceError: Failed to run swtpm daemon: /tmp/aexpect_XbFgBaCh/aexpect-u73lnw4a.sh: line 1: /usr/bin/swtpm: No such file or directory

LGTM

**Comment by xiagao** [comment]
- (007/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_on.flock_on.lock_posix_on.q35: PASS (97.46 s)
 (008/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_on.flock_off.lock_posix_off.q35: STARTED
\ (008/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_on.flock_off.lock_posix_off.q35: PASS (108.29 s)
 (009/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_on.flock_off.lock_posix_on.q35: STARTED
- (009/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_on.flock_off.lock_posix_on.q35: PASS (110.17 s)
 (010/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_off.flock_on.lock_posix_off.q35: STARTED
\ (010/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_off.flock_on.lock_posix_off.q35: PASS (91.98 s)
 (011/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_off.flock_on.lock_posix_on.q35: STARTED
/ (011/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_off.flock_on.lock_posix_on.q35: PASS (81.52 s)
 (012/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_off.flock_off.lock_posix_off.q35: STARTED
| (012/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_off.flock_off.lock_posix_off.q35: PASS (94.31 s)
 (013/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_off.flock_off.lock_posix_on.q35: STARTED
\ (013/104) Host_RHEL.m.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.x.0.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_off.flock_off.lock_posix_on.q35: PASS (98.33 s)

LGTM

File: virttest/qemu_devices/qdevices.py [hunk 1]
```diff
@@ -1773,6 +1773,9 @@ def __init__(self, name, aobject, child_bus=None):
         self.set_param('name', name)
         self._daemon_process = None
 
+    def daemon_process(self):
```
Comment (YongxueHong) [inline virttest/qemu_devices/qdevices.py:1776]:
```suggestion
    @property
    def daemon_process(self):
```

File: virttest/qemu_devices/qdevices.py [hunk 2]
```diff
@@ -1910,6 +1911,12 @@ def start_daemon(self):
         self.set_param('run_bg_kwargs', {'output_func': self._handle_log,
                                          'auto_close': False})
         super(QVirtioFSDev, self).start_daemon()
+        if not self.is_daemon_alive():
+            output = self.daemon_process().get_output()
+            self.daemon_process().close()
```
Comment (YongxueHong) [inline virttest/qemu_devices/qdevices.py:1916]: Avoid leaking, we should reset the `self._daemon_process = None`.
So I suggest adding a method to do this in the base class:

```
def close_daemon_process(self):
    self._daemon_process.close()
    self._daemon_process = None    
```

```suggestion
            self.close_daemon_process()
```

File: virttest/qemu_devices/qdevices.py (hunk 1 above)
Comment (YongxueHong) [reply virttest/qemu_devices/qdevices.py:1776]: Plus add the docstring about this method.

Comment (zhenyzha) [reply virttest/qemu_devices/qdevices.py:1777]: Thanks a lot, corrected.

File: virttest/qemu_devices/qdevices.py (hunk 2 above)
Comment (zhenyzha) [reply virttest/qemu_devices/qdevices.py]: Thanks a lot, corrected.

**Comment by zhenyzha** [comment]
git am 3348.patch
Applying: qdevices: Add daemon alive check in start_daemon()

**normal test:**
(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.arm64-pci: PASS (128.61 s)

’None -d params’ 
(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.aarch64.io-github-autotest-qemu.tpm_verify_device.with_emulator.arm64-pci: PASS (85.76 s)
‘Add -d params’
(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.aarch64.io-github-autotest-qemu.tpm_verify_device.with_emulator.arm64-pci: PASS (90.52 s)

**negative test:**
(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.arm64-pci: ERROR: VM 'avocado-vt-vm1' failed to start: Error occurred while executing make_create_command(). Check the log for traceback. (6.16 s)
virttest.qemu_devices.utils.DeviceError: Failed to run virtiofs daemon: /bin/bash: line 1: /usr/libexec/virtiofsd: No such file or directory

(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.aarch64.io-github-autotest-qemu.tpm_verify_device.with_emulator.arm64-pci: ERROR: VM 'avocado-vt-vm1' failed to start: Error occurred while executing make_create_command(). Check the log for traceback. (6.59 s)
virttest.qemu_devices.utils.DeviceError: Failed to run swtpm daemon: /tmp/aexpect_eswWziXu/aexpect-p46qcza_.sh: line 1: /usr/bin/swtpm: No such file or directory

@YongxueHong 
Could you pls help review when you are free? thanks.

File: virttest/qemu_devices/qdevices.py [hunk 3]
```diff
@@ -1965,6 +1972,13 @@ def start_daemon(self):
         self.set_param('cmd', tpm_cmd)
         self.set_param('run_bg_kwargs', {'auto_close': False})
         super(QSwtpmDev, self).start_daemon()
+        if not self.is_daemon_alive():
+            if self.daemon_process().get_status() != 0:
```
[the remark this answers is not in the document]
Comment (zhenyzha) [reply virttest/qemu_devices/qdevices.py]: Thanks a lot, corrected.

File: virttest/qemu_devices/qdevices.py [hunk 4]
```diff
@@ -1815,6 +1818,11 @@ def stop_daemon(self):
                 self._daemon_process.close()
                 self._daemon_process = None
```
Comment (YongxueHong) [inline virttest/qemu_devices/qdevices.py:1819]:
```suggestion
                self.close_daemon_process()
```

File: virttest/qemu_devices/qdevices.py [hunk 5]
```diff
@@ -1815,6 +1818,11 @@ def stop_daemon(self):
...
 
+    def close_daemon_process(self):
+        """Close daemon."""
```
Comment (YongxueHong) [inline virttest/qemu_devices/qdevices.py:1822]:
```suggestion
        """Close daemon process."""
```

File: virttest/qemu_devices/qdevices.py (hunk 4 above)
Comment (zhenyzha) [reply virttest/qemu_devices/qdevices.py]: corrected

File: virttest/qemu_devices/qdevices.py (hunk 5 above)
Comment (zhenyzha) [reply virttest/qemu_devices/qdevices.py]: corrected

**Comment by zhenyzha** [comment]
git am 3348.patch
Applying: qdevices: Add daemon alive check in start_daemon()

(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.arm64-pci: PASS (130.31 s)

@YongxueHong
Could you pls help review when you are free? thanks.

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Comment by qcheng-redhat** [comment]
Tested with latest code:

 (1/4) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: STARTED
| (1/4) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: PASS (121.99 s)
 (2/4) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: STARTED
| (2/4) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: PASS (103.45 s)
 (3/4) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: STARTED
| (3/4) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: PASS (187.90 s)
 (4/4) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: STARTED
\ (4/4) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: PASS (250.88 s)

**Comment by xiagao** [comment]
Run a test loop,the result is good.
LGTM
088-Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.X0.x86_64.io-github-autotest-qemu.virtio_fs_share_data_hugepage.basic_test.with_cache.none.q35 Finshed 2022-03-03 08:21:50 2022-03-03 08:23:34 PASS 103.355575
089-Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.X.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data_hugepage.with_numa.basic_test.with_cache.auto.q35 Finshed 2022-03-03 08:23:34 2022-03-03 08:25:27 PASS 113.126077
090-Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.X.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data_hugepage.with_numa.basic_test.with_cache.always.q35 Finshed 2022-03-03 08:25:27 2022-03-03 08:27:07 PASS 100.119214
091-Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.X.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data_hugepage.with_numa.basic_test.with_cache.none.q35 Finshed 2022-03-03 08:27:07 2022-03-03 08:28:51 PASS 104.287732
092-Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.X.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.run_stress.with_fio.with_cache.auto.q35 Finshed 2022-03-03 08:28:51 2022-03-03 08:35:15 PASS 384.115253
093-Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.run_stress.with_fio.with_cache.auto.with_writeback.q35 Finshed 2022-03-03 08:35:15 2022-03-03 09:01:07 PASS 1551.942066
094-Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.run_stress.with_fio.with_cache.always.q35 Finshed 2022-03-03 09:01:07 2022-03-03 09:08:02 PASS 414.819102
095-Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.x86_64.io-github-autotest-qemu.virtio_fs_share_data.run_stress.with_fio.with_cache.always.with_writeback.q35 Finshed 2022-03-03 09:08:02 2022-03-03 09:33:00 PASS 1497.431653

**Comment by YongxueHong** [comment]
Thanks all. Let's merge it.

**Pull request merged**